algorithms/linfa-clustering/src/k_means/hyperparameters.rs

```diff
@@ -123,81 +51,130 @@ impl<F: Float, R: Rng + Clone> KMeansHyperParams<F, R> {
     /// `n_clusters` is mandatory.
     ///
     /// Defaults are provided if optional parameters are not specified:
-    /// * `tolerance = 1e-4`;
-    /// * `max_n_iterations = 300`.
+    /// * `tolerance = 1e-4`
+    /// * `max_n_iterations = 300`
+    /// * `n_runs = 10`
+    /// * `init = KMeansPlusPlus`
 
-    pub fn new_with_rng(n_clusters: usize, rng: R) -> KMeansHyperParamsBuilder<F, R> {
-        KMeansHyperParamsBuilder {
+    pub fn new(n_clusters: usize, rng: R, dist_fn: D) -> Self {
+        if n_clusters == 0 {
```

Comment:
In previous implementations I returned an error in the fitting function. I think it's good practice to avoid panics in most parts of the code

Reply:
In the previous version invalid hyperparams triggered panics, so I did the same for `n_clusters`. Should I throw errors for every invalid hyperparam instead of panicking?

Reply:
If we really wanted to prevent illegal hyperparams without panicking we can use something like [`tightness`](https://crates.io/crates/tightness) to express constraints on individual hyperparams at the type level.

Reply:
I just added a short description and collected all errors into a single variant